yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+
+  /**
+   * Window limit in milliseconds
+   */
+  private val windowLimitMs = sparkConf.getLong(YarnHistoryProvider.OPTION_WINDOW_LIMIT,
+    YarnHistoryProvider.DEFAULT_WINDOW_LIMIT) * 1000
+
+  /**
+   * Number of events to get
+   */
+  private val eventFetchLimit = sparkConf.getLong(YarnHistoryProvider.OPTION_EVENT_FETCH_LIMIT,
+    YarnHistoryProvider.DEFAULT_EVENT_FETCH_LIMIT)
+
+  private val eventFetchOption: Option[Long] = if (eventFetchLimit > 0) Some(eventFetchLimit) else None
+
+  /**
+   * Start time. Doesn't use the `now` call as tests can subclass that and
+   * it won't be valid until after the subclass has been constructed
+   */
+  val serviceStartTime = System.currentTimeMillis()
+
+  /**
+   * Timeline endpoint URI
+   */
+  protected val timelineEndpoint = createTimelineEndpoint()
+
+  /**
+   * The timeline query client which uses the `jersey`
+   * Jersey instance to talk to a timeline service running
+   * at [[timelineEndpoint]], and creates a timeline (write) client instance
+   * to handle token renewal
+   *
+   */
+  protected val timelineQueryClient = {
+    createTimelineQueryClient()
+  }
+
+
+  /**
+   * Override point: create the timeline endpoint
+   * @return a URI to the timeline web service
+   */
+  protected def createTimelineEndpoint(): URI = {
+    getTimelineEndpoint(yarnConf)
+  }
+
+  /**
+   * Override point: create the timeline query client.
+   * This is called during instance creation.
+   * @return a timeline query client ot use for the duration
+   *         of this instance
+   */
+  protected def createTimelineQueryClient(): TimelineQueryClient = {
+    new TimelineQueryClient(timelineEndpoint, yarnConf, JerseyBinding.createClientConfig())
+  }
+
+  /**
+   * The empty listing, with a timestamp to indicate that the listing
+   * has never taken place.
+   */
+  private val emptyListing = new ApplicationListingResults(0, Nil, None)
+
+  /**
+   * List of applications. Initial result is empty
+   */
+  private var applications: ApplicationListingResults = emptyListing
+  
+  /**
+   * Last exception seen and when
+   */
+  protected var lastFailureCause: Option[(Throwable, Date)] = None
+
+  private val refreshCount = new AtomicLong(0)
+  private val refreshFailedCount = new AtomicLong(0)
+
+  /**
+   * Health marker
+   */
+  private val healthy = new AtomicBoolean(false)
+
+  /**
+   * Enabled flag
+   */
+  private val _enabled = timelineServiceEnabled(yarnConf)
+
+  /**
+   * Atomic boolean used to signal to the refresh thread that it
+   * must exit its loop.
+   */
+  private val stopRefresh = new AtomicBoolean(false)
+
+  /**
+   * refresher
+   */
+  val refresher = new Refresher()
+
+  /**
+   * Initialize the provider
+   */
+  init()
+
+  /**
+   * Check the configuration and log whether or not it is enabled;
+   * if it is enabled then the URL is logged too.
+   */
+  private def init(): Unit = {
+    if (!enabled) {
+      logError(YarnHistoryProvider.TEXT_SERVICE_DISABLED)
```

Comment:
I think it's a little weird for tests to depend on particular error messages, especially log messages. I never see logs as some sort of contract that tests should look at. Having constants also makes it hard to follow the code - I have to go look somewhere else for whatever it is that is being logged. If the log message contains parameters, it becomes even worse.